### PR TITLE
1816 expired link are not highlighted

### LINKED ISSUE #1815: 🔧 [Refactor] Properly order TransactionLink buttons
## 🔧 Refactor ticket

image
image
image
image

The buttons should always be next to each other with a fixed distance. Do not use two cols to separate them

### LINKED ISSUE #1816: 🐛 [Bug] Expired link are not highlighted
## 🐛 Bugreport

image

Expired link are not highlighted and still provide a copy and QR-Code button.

- provide the text "expired at" instead of "valid until"
- Highlight the expired link
- remove QR-Code and copy button for expired links

### PR DESCRIPTION
**Pull request by ogerly**
## 🍰 Pullrequest
- Transaction link list
1. removed button and added dropdown menu. for space reasons this is the best variant.
2. expired links only have a delete event and are somewhat paler than active links. 

### Issues
- fixes #1816 
- fixed #1815 

 
FireShot Capture 1157 -  - localhost

FireShot Capture 1156 -  - localhost

FireShot Capture 1155 -  - localhost

### CODE REVIEW

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 1]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
       expect(wrapper.find('div.transaction-link').exists()).toBeTruthy()
     })
 
-    describe('Copy link to Clipboard', () => {
-      const navigatorClipboard = navigator.clipboard
-      beforeAll(() => {
-        delete navigator.clipboard
-        navigator.clipboard = { writeText: navigatorClipboardMock }
+    describe('Link validUntil Date is not valid', () => {
+      it('find no link copy button', () => {
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:48]:
```suggestion
      it('has no copy link button', () => {
```

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 2]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
...
+        expect(wrapper.find('.test-copy-link').exists()).toBe(false)
       })
-      afterAll(() => {
-        navigator.clipboard = navigatorClipboard
+
+      it('find no Qr-Code Button ', () => {
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:52]:
```suggestion
      it('has no Qr-Code Button ', () => {
```

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 3]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
...
-        })
-        it('should call clipboard.writeText', () => {
-          expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
-            'http://localhost/redeem/c00000000c000000c0000',
-          )
-        })
-        it('toasts success message', () => {
-          expect(toastSuccessSpy).toBeCalledWith('gdd_per_link.link-copied')
-        })
+      it('find only Link Löschen Button ', () => {
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:56]:
```suggestion
      it('has  delete link button ', () => {
```

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 4]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
...
+        expect(wrapper.find('.test-delete-link').exists()).toBe(true)
       })
     })
 
-    describe('qr code modal', () => {
-      let spy
-
-      beforeEach(() => {
+    describe('Link validUntil Date is valid ', () => {
+      beforeEach(async () => {
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:62]:
```suggestion
      beforeEach(async () => {
        const now = new Date()
```

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 5]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
...
         jest.clearAllMocks()
+        await wrapper.setProps({
+          validUntil: '2022-05-30T14:22:40.000Z',
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:65]:
```suggestion
          validUntil: new Date(now.getFullYear(), now.getMonth(), now.getDate() + 2),
```

Comment (Mogge) [reply frontend/src/components/TransactionLinks/TransactionLink.spec.js:65]: Bei deiner Lösung funktioniert der Test ab dem 31. Mai nicht mehr

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 6]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
...
-          expect(wrapper.emitted('reset-transaction-link-list')).toBeTruthy()
+        describe('with success', () => {
+          beforeEach(async () => {
+            spy = jest.spyOn(wrapper.vm.$bvModal, 'show')
+            // spy.mockImplementation(() => Promise.resolve('some value'))
+            // mockAPIcall.mockResolvedValue()
+            await wrapper.find('.test-qr-code .dropdown-item').trigger('click')
+          })
+
+          it('qr-code Modal if show', () => {
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:111]:
```suggestion
          it('opens theqr-code Modal', () => {
```

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 7]
```diff
@@ -44,115 +44,139 @@ describe('TransactionLink', () => {
...
-          expect(toastErrorSpy).toBeCalledWith('Something went wrong :(')
+        describe('with success', () => {
+          beforeEach(async () => {
+            spy = jest.spyOn(wrapper.vm.$bvModal, 'msgBoxConfirm')
+            spy.mockImplementation(() => Promise.resolve('some value'))
+            mockAPIcall.mockResolvedValue()
+            await wrapper.find('.test-delete-link .dropdown-item').trigger('click')
+          })
+
+          it('test Modal if confirm true', () => {
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:132]:
```suggestion
          it('opens the modal ', () => {
```

**Review by Mogge — CHANGES_REQUESTED** [review]

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js (hunk 5 above)
Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.spec.js:65]: ok super, ich hatte vergeblich in nächtlichen versuchen ein new Date() erstellen wollen .. dann hatte ich es hart eingeschrieben.

**Comment by ogerly** [comment]
FireShot Capture 1155 -  - localhost

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 8]
```diff
@@ -1,49 +1,43 @@
...
+              <b-dropdown-item
+                v-if="validLink"
+                @click="$bvModal.show('modalPopover-' + id)"
+                class="pt-3 test-qr-code"
+              >
+                <b-img src="img/svg/qr-code.svg" width="18" class="filter"></b-img>
+                {{ $t('qrCode') }}
+              </b-dropdown-item>
+              <b-dropdown-item class="pt-3 test-delete-link" @click="deleteLink()">
+                <b-icon icon="trash" font-scale="1"></b-icon>
```
Comment (ulfgebhardt) [inline frontend/src/components/TransactionLinks/TransactionLink.vue:34]:
```suggestion
                <b-icon icon="trash"></b-icon>
```

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 9]
```diff
@@ -1,49 +1,43 @@
...
-              <b-img src="img/svg/qr-code.svg" width="60" class="filter"></b-img>
-            </b-button>
+          <b-col cols="12" lg="1" md="1" class="text-center text-md-right pr-5 pr-lg-4">
+            <b-dropdown no-caret right aria-expanded="false" size="sm">
+              <template #button-content>
+                <b-icon icon="three-dots-vertical"></b-icon>
+              </template>
+
+              <b-dropdown-item v-if="validLink" class="test-copy-link" @click="copy">
+                <b-icon icon="clipboard" font-scale="1"></b-icon>
```
Comment (ulfgebhardt) [inline frontend/src/components/TransactionLinks/TransactionLink.vue:22]:
```suggestion
                <b-icon icon="clipboard"></b-icon>
```

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 10]
```diff
@@ -115,6 +109,9 @@ export default {
     decay() {
       return `${this.amount - this.holdAvailableAmount}`
     },
+    validLink() {
+      return new Date(this.validUntil) > new Date()
```
Comment (ulfgebhardt) [inline frontend/src/components/TransactionLinks/TransactionLink.vue:113]: This I am not sure of, but consider this:

```suggestion
      return (new Date(this.validUntil)).getTime() > (new Date()).getTime()
```

File: frontend/src/components/TransactionLinks/TransactionLink.vue (hunk 9 above)
Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.vue:22]: changed

File: frontend/src/components/TransactionLinks/TransactionLink.vue (hunk 8 above)
Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.vue]: changed

File: frontend/src/components/TransactionLinks/TransactionLink.vue (hunk 10 above)
Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.vue:118]: no

**Comment by ulfgebhardt** [comment]
image
The elements are not properly hidden, the popup still contains a distance most likely there because of the hidden elements

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**